pkg/controller/backup/backup_controller.go

```diff
@@ -0,0 +1,119 @@
+func Add(mgr manager.Manager) error {
+	return AddToManager(mgr, NewReconciler(mgr))
+}
+
+// NewReconciler returns a new reconcile.Reconciler
+func NewReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileBackup{
+		Client: controllerutils.NewClientWithMetricsOrDie(mgr, controllerName),
+		scheme: mgr.GetScheme(),
+		logger: log.WithField("controller", controllerName),
+	}
+}
+
+// AddToManager adds a new Controller to mgr with r as the reconcile.Reconciler
+func AddToManager(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New(controllerName+"-controller", mgr, controller.Options{Reconciler: r, MaxConcurrentReconciles: controllerutils.GetConcurrentReconciles()})
+	if err != nil {
+		return err
+	}
+
+	reconciler := r.(*ReconcileBackup)
+
+	clusterDeploymentBackup := newClusterDeploymentBackup(reconciler)
+	err = clusterDeploymentBackup.registerClusterDeploymentWatch(c)
+	if err != nil {
+		return err
+	}
+
+	return nil // If we got here, then no errors occurred.
+}
+
+// This ensures that ReconcileBackup struct implements all functions that the reconcile.Reconciler interface requires.
+var _ reconcile.Reconciler = &ReconcileBackup{}
+
+// ReconcileBackup ensures that Velero backup objects are created when changes are made to Hive objects.
+type ReconcileBackup struct {
+	client.Client
+	scheme *runtime.Scheme
+
+	logger log.FieldLogger
+}
+
+// Reconcile ensures that all Hive object changes have corresponding Velero backup objects.
+// ClusterDeployment.Spec.Config.Machines
+// +kubebuilder:rbac:groups=hive.openshift.io,resources=clusterdeployments,verbs=get;list;watch
+// +kubebuilder:rbac:groups=velero.io,resources=backups,verbs=create
+func (r *ReconcileBackup) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	start := time.Now()
+	contextLogger := addReconcileRequestLoggerFields(r.logger, request)
+
+	// For logging, we need to see when the reconciliation loop starts and ends.
+	contextLogger.Info("reconciling backups and Hive object changes")
+	defer func() {
+		dur := time.Since(start)
+		hivemetrics.MetricControllerReconcileTime.WithLabelValues(controllerName).Observe(dur.Seconds())
+		contextLogger.WithField("elapsed", dur).Info("reconcile complete")
+	}()
+
+	clusterDeploymentBackup := newClusterDeploymentBackup(r)
+
+	// Get all ClusterDeployments that have a different hash than the last backup hash.
+	changedCDs, err := clusterDeploymentBackup.getChangedInNamespace(request.Namespace)
+	if err != nil {
+		contextLogger.WithError(err).Error("error finding changed cluster deployments in namespace")
+		return reconcile.Result{}, err
+	}
+
+	if len(changedCDs) == 0 {
+		// Nothing changed, so nothing to back up (i.e. don't create a Velero backup object).
+		return reconcile.Result{}, nil
+	}
+
+	// There are changes that need to be backed up.
+	err = controllerutils.CreateVeleroBackupObjectForNamespace(r.Client, request.Namespace)
```

Comment:
Just `CreateVeleroBackupObject` as function name sounds better. As the code passes the namespace as argument it is easy to understand it is per namespace.

Reply:
+1